### DISCUSSION TITLE
How to add turbine labels to a farm layout visualization in FLORIS?

### DISCUSSION THREAD

**Discussion by JackDotI** [NatLabRockies/floris#703] [Q&A] [answered]
Hi,

I posted this question on StackOverflow and I was asked by Raf to add the question here too. Worth adding I am new to FLORIS and fairly new to Python too!

**This is the original post:**
I am trying to add the turbine labels to the wind farm turbine plot created using FLORIS, but I am getting an error using the add_turbine_id_labels function. I have added an image of the plot I want to add turbine labels to. The line of code used to create the plot is:visualization.plot_turbines_with_fi(fi=fi) Turbine Plot
![image](https://github.com/NREL/floris/assets/117765122/0ec7b30a-0517-4a55-aa8b-a1a94877311d)

This is the error I am getting:

```
visualization.plot_turbines_with_fi(fi=fi)
visualization.add_turbine_id_labels(fi=fi, ax=plt.Axes)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[32], line 2
      1 visualization.plot_turbines_with_fi(fi=fi)
----> 2 visualization.add_turbine_id_labels(fi=fi, ax=plt.Axes)

File ~\anaconda3\lib\site-packages\floris\tools\visualization.py:140, in add_turbine_id_labels(fi, ax, **kwargs)
    134 layout_x, layout_y, _, _, _ = rotate_coordinates_rel_west(
    135     np.array([wind_direction]),
    136     coordinates_array
    137 )
    139 for i in range(fi.floris.farm.n_turbines):
--> 140     ax.annotate(
    141         i,
    142         (layout_x[0,0,i], layout_y[0,0,i]),
    143         xytext=(0,10),
    144         textcoords="offset points",
    145         **kwargs
    146     )

TypeError: annotate() missing 1 required positional argument: 'xy'
```

Please could someone let me know what I need to do to solve this error!

**I was given a solution following my initial post:**
```
fig, ax = plt.subplots()
visualization.plot_turbines_with_fi(fi=fi, ax=ax)
visualization.add_turbine_id_labels(fi=fi, ax=ax)
```
Which I am able to apply to the simple visualization above, however I am unsure how to apply this to the visualize_cut_plane visualization. 

Thank you for your help,
Jack

**Comment by paulf81** [comment, marked as answer]
Hi Jack,

  I put together an example mixing together some code from the examples 02_visualizations.py, 23_visualize_layout.py:

```
import matplotlib.pyplot as plt

from floris.tools import FlorisInterface
from floris.tools.layout_functions import visualize_layout

import floris.tools.visualization as wakeviz


# Declare a FLORIS interface
fi = FlorisInterface("inputs/gch.yaml")
fi.reinitialize(layout_x=[0, 100, 500, 1000, 1200,500], layout_y=[0, 800, 150, 500, 0,500])
fi.calculate_wake()

# Give turbines specific names
turbine_names = ['T01', 'T02','T03','S01','X01', 'X02']

# Plot the flow
fig, ax = plt.subplots(figsize=(8,8))
horizontal_plane = fi.calculate_horizontal_plane(
    x_resolution=200,
    y_resolution=100,
    height=90.0
)
wakeviz.visualize_cut_plane(horizontal_plane, ax=ax, title="Horizontal")

# Plot the turnine names
visualize_layout(
    fi,
    ax=ax,
    show_wake_lines=False,
    turbine_names=turbine_names
)

plt.show()
```

When I run this I get:

![image](https://github.com/NREL/floris/assets/6117702/7c2b7f06-8e15-4be8-97e5-59c333053685)

Does that work for you?

**Reply by rafmudaf** [reply]
@JackDotI they key here is that you hang on to one `Axes` object and pass it to each of the functions that you want to modify the `Axes`. In @paulf81's code above, the `ax` variable is returned from `plot.subplots()`. Then, it is passed in to the other FLORIS functions that will operate (in this case, add to) `ax`.

**Reply by JackDotI** [reply]
@paulf81 @rafmudaf Thank you both this has been really helpful! I have now got the visualizations working with the label.

**Reply by paulf81** [reply]
Glad to hear it!
